extensions/common/iam/oauth2/oauth2-client/src/main/java/org/eclipse/edc/iam/oauth2/Oauth2ClientExtension.java

```diff
@@ -0,0 +1,43 @@
+/*
+ *  Copyright (c) 2022 Amadeus
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Amadeus - initial API and implementation
+ *
+ */
+
+package org.eclipse.edc.iam.oauth2;
+
+import org.eclipse.edc.iam.oauth2.client.Oauth2ClientImpl;
+import org.eclipse.edc.iam.oauth2.spi.client.Oauth2Client;
+import org.eclipse.edc.runtime.metamodel.annotation.Extension;
+import org.eclipse.edc.runtime.metamodel.annotation.Inject;
+import org.eclipse.edc.runtime.metamodel.annotation.Provider;
+import org.eclipse.edc.spi.http.EdcHttpClient;
+import org.eclipse.edc.spi.system.ServiceExtension;
+import org.eclipse.edc.spi.system.ServiceExtensionContext;
+
+@Extension(value = Oauth2ClientExtension.NAME)
+public class Oauth2ClientExtension implements ServiceExtension {
+
+    public static final String NAME = "OAuth2 Core";
+
+    @Inject
+    private EdcHttpClient httpClient;
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Provider
```

Comment:
I have a proposal, because this is good as we provide extensibility, but the downside here is then when I add a `provision-oauth2` dependency then I'd have to explicitly add the `oauth2-client` as well. Couldn't the `oauth2-client` added in the build file of `provision-oauth2` (it is already on the `oauth2-core`) and this service provided as `isDefault = true`?
This way it won't be needed to have to explicitly define the client dependency (as it was done in the `control-plane` runtime of the end to end tests) but it would be extensible if needed.

Reply:
I am pretty sure that @paullatzelsperger would scream on it :D 
As default provider are by design not intended to be used for extensibility purposes as mentioned in the doc.

Reply:
As @bscholtes1A indicated, the `Oauth2ClientImpl` doesn't really fit the bill for a default provider, because it is not a "fallback" implementation, as in: "if there is nothing else available, use this".
In most cases people won't write their own `Oauth2Client` because the one we provide is deemed production-ready.
That means we _could_ let the `provision-oauth2` module explicitly depend on `oauth2-client`, presumably as `runtimeOnly`. That will mean however, that if someone wants to indeed write their own `Oauth2Client` implementation, they'd have to explicitly exclude that (transitive) dependency onto `oauth2-client` from their launcher config, otherwise they'd end up with multiple `Oauth2Client` impls on the classpath, which is undefined behaviour.

How about instead of doing that we provide a BOM that pulls in all of the OAuth2 modules? That way it would satisfy the needs of 99%, while still providing flexibility for the wicked :)

Reply:
> How about instead of doing that we provide a BOM that pulls in all the OAuth2 modules? That way people could just use that, and it would satisfy the needs of 99%, while still providing flexibility for the wicked :)

What if `oauth2` is used for provisioning but not for authenticating on the network? that BOM could be misleading because it will register an `IdentityService` instance. 
An `oauth2-client` is something that is likely needed by every `oauth2` extension, but having a BOM that loads every `oauth2`-related extension doesn't seems to be something really usable.

Reply:
that could be a separate BOM then: we could have an "OauthAuthentication" and an "OAuthProvisioning" BOM. Using both these BOMs should not cause any problems.

I think we should not declare that dependency on the module/artifact level, if we foresee the slightest chance that anyone will write their own `Oauth2Client` implementation. 

If we don't, then declaring it on the module/artifact level is fine, but in that case I would go as far as to say (and document) that provisioning with `OAuth2` is only supported with the `OAuth2ClientImpl`, everything else is a workaround.

Reply:
@paullatzelsperger where would you see these BOM located in the codebase?

Reply:
@bscholtes1A for our purposes here, a BOM is just a gradle build script that aggregates dependencies (i.e. the other modules), so they would probably have to be in separate modules e.g. `oauth-core-bom` and `provision-oauth-bom` that contain only the aforementioned build files.

Reply:
Similar to Paul, I think coupling the modules via dependency is not the best way to handle this. If they are coupled, they should be contained in a common module. However, that's not the case here so BOMs seem the right mechanism to use.

Reply:
To clarify, by dependency in the module, I meant two “code” modules. That is a smell. If one module depends on another implementation that is not a library, then there should be one module; otherwise use a BOM.

Reply:
@jimmarino  @paullatzelsperger @ndr-brt 
I created the two BOMs `oauth2-service-bom` and `provision-oauth2-bom` in the last commit. 
Could you please check if that looks better now?